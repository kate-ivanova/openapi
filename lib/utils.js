"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const prettier_1 = __importDefault(require("prettier"));
const http = __importStar(require("http"));
const fs = __importStar(require("fs"));
const generate_1 = require("./generate");
function readByPath(path) {
    return new Promise((resolve, reject) => {
        fs.readFile(path, 'utf-8', (err, file) => {
            if (err) {
                reject(err);
            }
            else {
                resolve(file);
            }
        });
    });
}
exports.readByPath = readByPath;
function readByUrl(url) {
    return new Promise((resolve, reject) => {
        http.get(url, (res) => {
            const data = [];
            res.on('data', (chunk) => {
                data.push(chunk);
            });
            res.on('close', () => {
                resolve(data.join(''));
            });
            res.on('error', (err) => reject(err));
        });
    });
}
exports.readByUrl = readByUrl;
function transform(schemaData) {
    const docs = JSON.parse(schemaData);
    const code = generate_1.generate(docs);
    return prettier_1.default.format(code, {
        tabWidth: 2,
        semi: true,
        singleQuote: true,
        parser: 'typescript',
    });
}
exports.transform = transform;
//# sourceMappingURL=utils.js.map