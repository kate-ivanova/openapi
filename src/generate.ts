import type * as Swagger from 'swagger-schema-official';

import groupBy from 'lodash/groupBy';
import upperFirst from 'lodash/upperFirst';

type TKey = 'get' | 'put' | 'post' | 'delete' | 'options' | 'head' | 'patch';
const TMETHOD: TKey[] = [
  'get',
  'put',
  'post',
  'delete',
  'options',
  'head',
  'patch',
];

function getRefName(ref: string) {
  return ref.replace(/^#\/definitions\//, '');
}

function objToString(obj: any): string {
  const result: string[] = ['{'];
  Object.keys(obj).forEach((k) => {
    const val = obj[k];
    if (k[k.length - 1] === '?') {
      result.push(JSON.stringify(k.slice(0, k.length - 1)) + '?:');
    } else {
      result.push(JSON.stringify(k) + ':');
    }

    if (typeof val !== 'object') {
      result.push(`${val};`);
    } else {
      result.push(`${objToString(val)};`);
    }
  });
  return result.concat('}').join('');
}

/** Used to match property names within property paths. */
var rePropName = /[^.[\]]+|\[(?:(-?\d+(?:\.\d+)?)|(["'])((?:(?!\2)[^\\]|\\.)*?)\2)\]|(?=(?:\.|\[\])(?:\.|\[\]|$))/g;

/** Used to match backslashes in property paths. */
var reEscapeChar = /\\(\\)?/g;

/**
 * Converts `string` to a property path array.
 *
 * @private
 * @param {string} string The string to convert.
 * @returns {Array} Returns the property path array.
 */
const stringToPath = function (str: string) {
  var result = [];
  if (str.charCodeAt(0) === 46 /* . */) {
    result.push('');
  }
  str.replace(
    rePropName,
    (match: any, num: any, quote: any, subString: any) => {
      result.push(quote ? subString.replace(reEscapeChar, '$1') : num || match);
      return '';
    }
  );
  return result;
};

function generateProperties(schema: Swagger.Schema): string {
  const result: string[] = [];

  if (schema.$ref) {
    result.push(getRefName(schema.$ref));
  }

  switch (schema.type as string) {
    case 'object':
      {
        result.push('{');
        const props = schema.properties ?? {};
        Object.keys(props).forEach((key) => {
          const p = props[key];
          result.push(`${key}: ${generateProperties(p)};`);
        });
        result.push('}');
      }
      break;
    case 'array':
      {
        const items = schema.items;
        const itemType = Array.isArray(items)
          ? items.map((item) => generateProperties(item)).join(' | ')
          : items
          ? generateProperties(items)
          : 'unknown';
        result.push(`Array<${itemType}>`);
      }
      break;

    case 'string':
      if (schema.enum) {
        result.push(schema.enum.map((p) => JSON.stringify(p)).join('|'));
      } else {
        result.push('string');
      }
      break;

    case 'boolean':
    case 'Boolean':
      if (schema.enum) {
        result.push(schema.enum.join('|'));
      } else {
        result.push('boolean');
      }
      break;

    case 'number':
    case 'integer':
    case 'float':
      if (schema.enum) {
        result.push(schema.enum.join('|'));
      } else {
        result.push('number');
      }
      break;

    case 'file':
      result.push('IFileType$$');
    default:
      break;
  }

  return result.join('\n');
}

type TSchema = Swagger.BaseSchema & { name: string; required: boolean };

function transformInputSchema<T extends TSchema>(
  params: T[] | undefined,
  def = 'null'
) {
  if (!params || !params.length) {
    return def;
  }
  const obj = params.reduce((memo: any, param: T) => {
    const namePath = stringToPath(param.name);
    let ptr = memo;
    const value = generateProperties(param);
    for (let i = 0, iLen = namePath.length; i < iLen; i++) {
      if (i + 1 === iLen) {
        const name = param.required ? namePath[i] : `${namePath[i]}?`;
        ptr[name] = value;
      } else {
        ptr = ptr[namePath[i]] ?? (ptr[namePath[i]] = {});
      }
    }
    return memo;
  }, {} as any);
  return objToString(obj);
}

function generateMethod(requestName: string, operation: Swagger.Operation) {
  const result: string[] = [];
  if (operation.description) {
    result.push(`\n// ${operation.description}`);
  }

  const resultType =
    Object.keys(operation.responses)
      .reduce((memo, key) => {
        const item = operation.responses[key];

        const ref = item as Swagger.Reference;

        if (ref.$ref) {
          memo.push(getRefName(ref.$ref));
        } else {
          const res = item as Swagger.Response;

          if (res.schema) {
            const name = generateProperties(res.schema);
            memo.push(name);
          }
        }

        return memo;
      }, [] as string[])
      .join('|') || 'unknown';

  result.push(`get(param: ${requestName} & TOptions): Promise<${resultType}>;`);

  return result.join('\n');
}

function generateAPI(docs: Swagger.Spec) {
  const result: string[] = [];
  const typesResult: string[] = [];

  const paths = docs.paths;

  Object.keys(paths).forEach((url) => {
    const path = paths[url];
    TMETHOD.forEach((method) => {
      if (path[method] === undefined) {
        return;
      }
      const operation = path[method]!;

      const groups = groupBy<TSchema>(
        (operation.parameters?.filter((p: any) => p.in) as any) ?? [],
        (p: any) => p.in
      );

      const NameParam = upperFirst(operation.operationId!);
      const options = Object.keys(groups).reduce((memo, groupName) => {
        const val = transformInputSchema(groups[groupName]);
        if (val !== 'null') {
          const name = `I${upperFirst(groupName)}${NameParam}`;
          typesResult.push(`export interface ${name} ${val}`);
          memo[groupName] = name;
        }
        return memo;
      }, {} as Record<string, string>);

      const RequestName = `I${NameParam}Request`;
      const typeParams = objToString({
        method: JSON.stringify(method.toUpperCase()),
        url: JSON.stringify(url),
        ...options,
      });
      typesResult.push(`export interface ${RequestName} ${typeParams}`);

      const res = generateMethod(RequestName, path[method]!);
      result.push(res);
    });
  });

  result.push(
    `get({ path, ...options}: any): Promise<any> {`,
    `options.url = !path ? options.url : Object.keys(path).reduce((memo, key) => memo.replace(new RegExp('{' + key + '}'), path[key]), options.url);`,
    `return this.call(options);`,
    `}`
  );

  return [
    typesResult.join('\n'),
    'export type IFileType$$ = string;',
    'export interface IOptionsBase$$ { body?: any; query?: any; header?: any; formData?: any; }',
    'export abstract class API<TOptions = {}> {',
    'abstract call(param: any & TOptions): Promise<any>;\n',
    result.join('\n'),
    '}',
  ].join('\n');
}

export function generate(docs: Swagger.Spec) {
  const result: string[] = [];
  const definitions = docs.definitions;
  if (definitions) {
    Object.keys(definitions).forEach((def) => {
      const schema = definitions[def];
      const code = [
        schema.description ? `// ${schema.description}` : '',
        `export interface ${def}` + generateProperties(schema),
      ]
        .filter((p) => p)
        .join('\n');
      result.push(code);
    });
  }
  result.push(generateAPI(docs));

  return result.join('\n');
}
