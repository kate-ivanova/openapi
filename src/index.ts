import { readByUrl, readByPath, transform } from './utils';

if (process.argv.length !== 4) {
  process.exit(1);
}

const command: any = process.argv[2];
const filepath = process.argv[3];

const config: Partial<Record<string, (p: string) => Promise<string>>> = {
  ['--url'](filepath: string) {
    return readByUrl(filepath)
  },
  ['--file'](filepath: string) {
    return readByPath(filepath);
  }
}

const fn = config[command];
if (!fn) {
  process.exit(1);
}

fn(filepath).then(transform).then(console.log);
