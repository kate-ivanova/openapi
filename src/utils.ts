import prettier from 'prettier';
import * as http from 'http';
import type * as Swagger from 'swagger-schema-official';
import { generate } from './generate';

export function readByUrl(url: string) {
  return new Promise<any>((resolve, reject) => {
    http.get(url, (res) => {
      const data: string[] = [];
      res.on('data', (chunk) => {
        data.push(chunk);
      });
      res.on('close', () => {
        resolve(data.join(''));
      });
      res.on('error', (err) => reject(err));
    });
  });
}

export function transform(schemaData: string) {
    const docs: Swagger.Spec = JSON.parse(schemaData);
    const code = generate(docs);
    return prettier.format(code, {
      tabWidth: 2,
      semi: true,
      singleQuote: true,
      parser: 'typescript',
    });
}
