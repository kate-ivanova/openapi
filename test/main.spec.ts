import * as fs from 'fs';
import * as path from 'path';
import { transform } from '../src/utils';

describe('main', () => {
  let data: string;
  beforeAll(() => {
    const filePath = path.resolve(__dirname, 'fixture', 'schema.json');
    data = fs.readFileSync(filePath).toString('utf-8');
  });

  it('generate', async () => {
    const code = transform(data);
    expect(code).toMatchSnapshot();
    const filePath = path.resolve(__dirname, 'code.ts');

    fs.writeFileSync(filePath, code);

    const API = require(filePath).API;
    class Client extends API {
        obj: any = null;
        call(param: any) {
            this.obj = param;
            return Promise.resolve();
        }
    }
    const client = new Client();
    await client.get({
        method: 'GET',
        url: '/v1/offerings/{slug}',
        path: { slug: 'test' },
        query: { include: 'test' }
    });
    expect(client.obj).toMatchSnapshot();
  });
});
