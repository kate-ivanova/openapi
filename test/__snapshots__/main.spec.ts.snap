// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`openapi generate 1`] = `
"// Set user password
export interface AuthSerializer {
  data: {
    id: string;
    type: string;
    attributes: {
      first_name: string;
      last_name: string;
      slug: string;
      avatar_url: string;
      jwt_token: string;
      need_set_password: boolean;
    };
  };
}
// Get all offerings
export interface OfferingSerializer {
  data: {
    id: string;
    type: string;
    attributes: {
      company_name: string;
      state: string;
      pitch: string;
      max_raise_amount_cents: string;
      description: string;
      logo_url: string;
      cover_url: string;
      supported_by_ios_app: string;
      pitch_highlights: string;
      requires_w9: boolean;
      requires_alcohol_reps: boolean;
      reached_max_raise_amount: boolean;
      reached_min_raise_amount: boolean;
      min_investment_amount_cents: number;
      max_investment_amount_cents: number;
      min_raise_amount_cents: number;
      min_wire_amount_cents: number;
      investment_amount_multiplier_cents: number;
      investors_count: number;
      discussions_count: number;
      reviews_count: number;
      follows_count: number;
      updates_count: number;
      avatar_extra_small_url: string;
      avatar_extra_small_2x_url: string;
      avatar_small_url: string;
      avatar_small_2x_url: string;
      avatar_medium_url: string;
      avatar_medium_2x_url: string;
      avatar_large_url: string;
      avatar_large_2x_url: string;
      card_image_url: string;
      video_url: string;
      deadline: number;
      investment_change_deadline: number;
      card_enabled: boolean;
      published_at: number;
      amount_raised_cents: number;
      amount_raised_this_week_cents: number;
      custodial: boolean;
    };
    relationships: {
      security: {
        data: {
          id: string;
          type: string;
        };
      };
      perks: {
        data: Array<{
          id: string;
          type: string;
        }>;
      };
      faqs: {
        data: Array<{
          id: string;
          type: string;
        }>;
      };
      previous_campaigns: {
        data: Array<{
          id: string;
          type: string;
        }>;
      };
      investment_reasons: {
        data: Array<{
          id: string;
          type: string;
        }>;
      };
      documents: {
        data: Array<{
          id: string;
          type: string;
        }>;
      };
      risks: {
        data: Array<{
          id: string;
          type: string;
        }>;
      };
      questions: {
        data: Array<{
          id: string;
          type: string;
        }>;
      };
      press: {
        data: Array<{
          id: string;
          type: string;
        }>;
      };
      roles_team_members: {
        data: Array<{
          id: string;
          type: string;
        }>;
      };
      tags: {
        data: Array<{
          id: string;
          type: string;
        }>;
      };
      pitch_sections: {
        data: Array<{
          id: string;
          type: string;
        }>;
      };
      exit: {
        data: {
          id: string;
          type: string;
        };
      };
      form_c: {
        data: {
          id: string;
          type: string;
        };
      };
      issuer_profile: {
        data: {
          id: string;
          type: string;
        };
      };
    };
  };
}
export interface IPathPostV1AuthProviderOauth {
  provider: 'facebook' | 'twitter';
}
export interface IFormDataPostV1AuthProviderOauth {
  oauth_token: string;
  oauth_token_secret?: string;
  referral_token?: string;
}
export interface IPostV1AuthProviderOauthRequest {
  method: 'POST';
  url: '/v1/auth/{provider}/oauth';
  path: IPathPostV1AuthProviderOauth;
  formData: IFormDataPostV1AuthProviderOauth;
}
export interface IPathGetV1OfferingsSlug {
  slug: string;
}
export interface IQueryGetV1OfferingsSlug {
  include?: string;
}
export interface IGetV1OfferingsSlugRequest {
  method: 'GET';
  url: '/v1/offerings/{slug}';
  path: IPathGetV1OfferingsSlug;
  query: IQueryGetV1OfferingsSlug;
}
export type IFileType$$ = string;
export interface IOptionsBaseT<T> {
  body?: T;
  query?: T;
  header?: T;
  formData?: T;
  path?: T;
  url: string;
  method: string;
}
export abstract class API<TOptions = {}> {
  abstract call(param: IOptionsBaseT<any> & TOptions): Promise<any>;

  // auth with social networks
  request(
    param: IPostV1AuthProviderOauthRequest & TOptions
  ): Promise<AuthSerializer>;

  // Get Offering
  request(
    param: IGetV1OfferingsSlugRequest & TOptions
  ): Promise<OfferingSerializer>;
  request(params: IOptionsBaseT<{}> & TOptions): Promise<any> {
    const options: any = params;
    if (options.path) {
      options.url = Object.keys(options.path).reduce(
        (memo, key) =>
          memo.replace(new RegExp('{' + key + '}'), options.path[key]),
        options.url
      );
    }
    if (options.query) {
      const query = Object.keys(options.query).reduce((memo, key) => {
        memo.push(key + '=' + options.query[key]);
        return memo;
      }, [] as string[]);
      if (query.length) {
        options.url +=
          (options.url.indexOf('?') === -1 ? '?' : '') + query.join('&');
      }
    }
    return this.call(options);
  }
}
"
`;

exports[`openapi generate 2`] = `
Object {
  "method": "GET",
  "path": Object {
    "slug": "test",
  },
  "query": Object {
    "include": "1234",
  },
  "url": "/v1/offerings/test?include=1234",
}
`;
